### PR TITLE
SWC-6365

### PR DESCRIPTION
**Pull request by jay-hodgson**

### CODE REVIEW

File: packages/synapse-react-client/src/lib/containers/oauth/CreateOAuthClient.tsx [hunk 1]
```diff
@@ -64,17 +93,21 @@ export const CreateOAuthModal: React.FunctionComponent<
...
-    const propArr = client?.redirect_uris
-
-    client &&
-    (stateArr?.length !== propArr?.length ||
-      !stateArr?.every(el => propArr?.includes(el)) ||
-      sectorUri != client?.sector_identifier_uri)
-      ? setWarnTrigger(true)
-      : setWarnTrigger(false)
-  }, [redirectUris, sectorUri, client])
+  useDebouncedEffect(
```
Comment (jay-hodgson) [inline packages/synapse-react-client/src/lib/containers/oauth/CreateOAuthClient.tsx:96]: This is called when the client-side oAuthClient is updated from the UI, so I debounce the effect.

File: packages/synapse-react-client/src/lib/containers/oauth/CreateOAuthClient.tsx [hunk 2]
```diff
@@ -64,17 +93,21 @@ export const CreateOAuthModal: React.FunctionComponent<
...
+    () => {
+      if (accessToken) {
+        // SWC-6365: use the pre-check service to determine if we need to show a warning on edit
+        SynapseClient.isOAuthClientReverificationRequired(
+          getOAuthClient(),
+          accessToken,
+        ).then(precheckResult => {
+          setWarnTrigger(precheckResult.reverificationRequired)
+        })
```
Comment (jay-hodgson) [inline packages/synapse-react-client/src/lib/containers/oauth/CreateOAuthClient.tsx:105]: If there's an error in the precheck, I think the update will also fail (at which point the user will see the error).

File: packages/synapse-react-client/src/lib/utils/SynapseClient.ts [hunk 3]
```diff
@@ -2445,6 +2446,23 @@ export const updateOAuthClient = (
   )
 }
 
+/**
+ 
+ Note: Only the creator of a client can update it.
+ https://rest-docs.synapse.org/rest/PUT/oauth2/client/id/verificationPrecheck.html
+ */
+export const isOAuthClientReverificationRequired = (
```
Comment (jay-hodgson) [inline packages/synapse-react-client/src/lib/utils/SynapseClient.ts:2454]: New API call

File: packages/synapse-react-client/src/lib/containers/oauth/CreateOAuthClient.tsx [hunk 4]
```diff
@@ -53,6 +57,31 @@ export const CreateOAuthModal: React.FunctionComponent<
...
+  }, [
+    client?.client_id,
+    client?.etag,
+    clientName,
+    clientUri,
+    policyUri,
+    redirectUris,
+    sectorUri,
+    tosUri,
+  ])
```
Comment (nickgros) [inline packages/synapse-react-client/src/lib/containers/oauth/CreateOAuthClient.tsx:83]: This is being invoked every render, so can simplify to `useMemo`
```suggestion
  const oAuthClient: OAuthClient = useMemo(() => {
    return {
      client_id: client?.client_id,
      client_name: clientName,
      redirect_uris: redirectUris?.map(str => str.uri) ?? [''],
      policy_uri: policyUri,
      client_uri: clientUri,
      sector_identifier_uri: sectorUri ?? '',
      tos_uri: tosUri,
      etag: client?.etag,
    }
  }, [
    client?.client_id,
    client?.etag,
    clientName,
    clientUri,
    policyUri,
    redirectUris,
    sectorUri,
    tosUri,
  ])
```

File: packages/synapse-react-client/src/lib/containers/oauth/CreateOAuthClient.tsx [hunk 5]
```diff
@@ -64,17 +93,21 @@ export const CreateOAuthModal: React.FunctionComponent<
...
-      sectorUri != client?.sector_identifier_uri)
-      ? setWarnTrigger(true)
-      : setWarnTrigger(false)
-  }, [redirectUris, sectorUri, client])
+  useDebouncedEffect(
+    () => {
+      if (accessToken) {
+        // SWC-6365: use the pre-check service to determine if we need to show a warning on edit
+        SynapseClient.isOAuthClientReverificationRequired(
+          getOAuthClient(),
```
Comment (nickgros) [inline packages/synapse-react-client/src/lib/containers/oauth/CreateOAuthClient.tsx:101]: Following above suggestion
```suggestion
          oAuthClient,
```

File: packages/synapse-react-client/src/lib/containers/oauth/CreateOAuthClient.tsx (hunk 2 above)
Comment (nickgros) [reply packages/synapse-react-client/src/lib/containers/oauth/CreateOAuthClient.tsx:105]: Makes sense, we probably don't want to show an outright error until they click submit

File: packages/synapse-react-client/src/lib/containers/oauth/CreateOAuthClient.tsx [hunk 6]
```diff
@@ -128,16 +161,7 @@ export const CreateOAuthModal: React.FunctionComponent<
...
-          client_id: client?.client_id,
-          client_name: clientName,
-          redirect_uris: redirectUris?.map(str => str.uri) ?? [''],
-          policy_uri: policyUri,
-          client_uri: clientUri,
-          sector_identifier_uri: sectorUri ?? '',
-          tos_uri: tosUri,
-          etag: client?.etag,
-        }
+        const oAuthClient = getOAuthClient()
```
Comment (nickgros) [inline packages/synapse-react-client/src/lib/containers/oauth/CreateOAuthClient.tsx:164]: Following above suggestion
```suggestion
```

File: packages/synapse-react-client/src/lib/containers/oauth/CreateOAuthClient.tsx [hunk 7]
```diff
@@ -64,17 +93,21 @@ export const CreateOAuthModal: React.FunctionComponent<
...
+        // SWC-6365: use the pre-check service to determine if we need to show a warning on edit
+        SynapseClient.isOAuthClientReverificationRequired(
+          getOAuthClient(),
+          accessToken,
+        ).then(precheckResult => {
+          setWarnTrigger(precheckResult.reverificationRequired)
+        })
+      }
+    },
+    [accessToken, getOAuthClient],
```
Comment (nickgros) [inline packages/synapse-react-client/src/lib/containers/oauth/CreateOAuthClient.tsx:108]: Following above suggestion
```suggestion
    [accessToken, oAuthClient],
```

**Review by nickgros — APPROVED** [review]
LGTM, one suggested simplification

**Comment by jay-hodgson** [comment]
Memorized value (via useMemo()) rather than the function (useCallback()), as Nick recommended.

**Pull request merged**